src/Workspaces/SharedUtilitiesAndExtensions/Workspace/Core/CodeGeneration/Symbols/SymbolMappingFactory.cs

```diff
@@ -0,0 +1,638 @@
+                    {
+                        Contract.ThrowIfFalse(mappedCodeGenerationMethods.Contains(declaredCodeGenerationMethod), $"Expected to map code generation method '{declaredCodeGenerationMethod}'");
+                    }
+                }
+            }
+
+            var constructedType = typeBuilder.CreateTypeInfo()?.AsType();
+            Contract.ThrowIfNull(constructedType);
+            Contract.ThrowIfFalse(ImmutableInterlocked.TryAdd(ref _implementationTypes, baseType, constructedType));
+
+            return constructedType;
+        }
+    }
+
+    private static void LoadLocal(ILGenerator ilGenerator, LocalBuilder local)
+    {
+        switch (local.LocalIndex)
+        {
+            case 0:
+                ilGenerator.Emit(OpCodes.Ldloc_0);
+                break;
+            case 1:
+                ilGenerator.Emit(OpCodes.Ldloc_1);
+                break;
+            case 2:
+                ilGenerator.Emit(OpCodes.Ldloc_2);
+                break;
+            case 3:
+                ilGenerator.Emit(OpCodes.Ldloc_3);
+                break;
+            case <= byte.MaxValue:
+                ilGenerator.Emit(OpCodes.Ldloc_S, local);
+                break;
+            default:
+                ilGenerator.Emit(OpCodes.Ldloc, local);
+                break;
+        }
+    }
+
+    private static void LoadLocalAddress(ILGenerator ilGenerator, LocalBuilder local)
+    {
+        switch (local.LocalIndex)
+        {
+            case <= byte.MaxValue:
+                ilGenerator.Emit(OpCodes.Ldloca_S, local);
+                break;
+            default:
+                ilGenerator.Emit(OpCodes.Ldloca, local);
+                break;
+        }
+    }
+
+    private static void GenerateDefaultReturnForType(ILGenerator ilGenerator, Type type)
+    {
+        if (type != typeof(void))
+        {
+            // Types that do not currently have special cases, but may need them:
+            //
+            //  Task
+            //  Task<T>
+            //  ValueTask<T> where we want to return something other than default(T)
+            //  IEnumerable<T> and derived interfaces
+            //  Immutable collections, except for ImmutableArray<T> which returns empty
+            var genericTypeDefinition = type.IsGenericType ? type.GetGenericTypeDefinition() : type;
+            if (genericTypeDefinition == typeof(ImmutableArray<>))
+            {
+                // Return ImmutableArray<T>.Empty instead of default(ImmutableArray<T>)
+                var fieldInfo = type.GetField(nameof(ImmutableArray<int>.Empty), BindingFlags.Public | BindingFlags.Static);
+                Contract.ThrowIfNull(fieldInfo);
+                ilGenerator.Emit(OpCodes.Ldsfld, fieldInfo);
+            }
+            else if (genericTypeDefinition.IsValueType)
+            {
+                var local = ilGenerator.DeclareLocal(type);
```

Comment:
i'm not ok with this.  we should not be emitting IL at runtime to deal with the case of trying to talk to apis our own team publishes for us to consume.

Reply:
➡️ This change is part of prerequisite pull request https://github.com/dotnet/roslyn/pull/73675. Resolving this comment as off-topic.